### PR TITLE
[rollout, diffusion] refactor: integrate Qwen-Image step execution into FlowGRPO and MixGRPO

### PR DESCRIPTION
**Pull request by Sky-Trigger**
### What does this PR do?

This PR implements the first stage of #251 by promoting the existing Qwen-Image step-execution support from `verl_omni/experimental` into the existing FlowGRPO and MixGRPO pipeline adapters.

With this change, `step_execution=True` becomes an execution mode of the existing `flow_grpo` and `mix_grpo` algorithm registrations instead of requiring separate `flow_grpo_stepwise` and `mix_grpo_stepwise` registrations.

The original full-forward execution path remains available when `step_execution=False`.

### Motivation

The previous implementation maintained normal execution and step execution as separate pipeline registrations:

* `flow_grpo` and `flow_grpo_stepwise`
* `mix_grpo` and `mix_grpo_stepwise`

This duplicated the rollout control flow and required the experimental stepwise adapters to remain synchronized with changes to the normal pipeline implementations.

vLLM-Omni already provides a standard step-execution lifecycle:

* `prepare_encode`
* `denoise_step`
* `step_scheduler`
* `post_decode`

This PR integrates that lifecycle directly into the existing Qwen-Image FlowGRPO and MixGRPO adapters, making step execution a runtime option of the same algorithm implementation.

This reduces duplicated code, removes the need for parallel experimental registrations, and keeps the normal and step-execution paths under the same rollout output contract.

#### FlowGRPO

This PR integrates step-execution support into the existing `QwenImagePipelineWithLogProb` adapter.

The step-execution path now:

* accepts pre-tokenized prompt IDs and attention masks;
* initializes prompt embeddings, negative prompt embeddings, latents, timesteps, scheduler state, CFG state, and SDE parameters in `prepare_encode`;
* performs one transformer denoising pass in `denoise_step`;
* performs SDE scheduler updates and log-probability collection in `step_scheduler`;
* packages the final rollout trajectory in `post_decode`;
* moves returned tensors to CPU before inter-process transfer;
* preserves the regular full-forward path when step execution is disabled.

The step-execution path preserves the existing FlowGRPO rollout output fields:

* `all_latents`
* `all_log_probs`
* `all_timesteps`
* `prompt_embeds`
* `prompt_embeds_mask`
* `negative_prompt_embeds`
* `negative_prompt_embeds_mask`

The live request latents are kept in FP32 throughout step execution. The transformer input is still cast to the model dtype inside `denoise_step`.

This avoids mixed-dtype batches when new FP32 requests are added to a batch containing requests that have already completed one or more denoising steps.

#### MixGRPO

The MixGRPO adapter reuses the FlowGRPO step-execution implementation while preserving its own SDE-window selection behavior.

The following sampling strategies remain supported:

* `random`
* `progressive`

For the seeded random strategy, the SDE window is selected using:

```text
sde_window_seed + global_steps
```

This keeps the selected window consistent across rollout ranks for the same training step.

For the progressive strategy, the SDE window advances based on:

* `global_steps`
* `sde_window_size`
* `iters_per_group`

The same window initialization is now applied in both:

* the normal `forward()` path;
* the step-execution `prepare_encode()` path.

This ensures that enabling step execution does not change the MixGRPO sampling-window behavior.

#### Registration cleanup

The following experimental adapters are removed:

* `qwen_image_flow_grpo_stepwise`
* `qwen_image_mix_grpo_stepwise`

The automatic mapping from an algorithm name to its `<algorithm>_stepwise` variant is also removed.

When step execution is enabled, the rollout continues using the original algorithm registration:

* `flow_grpo`
* `mix_grpo`

The engine selects the step-execution lifecycle through the rollout configuration rather than through a separate algorithm name.

### Usage

Step execution can be enabled directly on the existing FlowGRPO or MixGRPO pipeline:

```bash
actor_rollout_ref.rollout.step_execution=True \
actor_rollout_ref.rollout.max_num_seqs=16
```

Equivalent YAML configuration:

```yaml
actor_rollout_ref:
  rollout:
    step_execution: true
    max_num_seqs: 16
```

`max_num_seqs=16` sets the maximum number of diffusion requests that can be scheduled concurrently during step execution.

No separate `flow_grpo_stepwise` or `mix_grpo_stepwise` registration is required.

### Validation

End-to-end Qwen-Image FlowGRPO training was run with the same training configuration under the following two settings.

Baseline:

```bash
actor_rollout_ref.rollout.step_execution=False
```

Step execution:

```bash
actor_rollout_ref.rollout.step_execution=True \
actor_rollout_ref.rollout.max_num_seqs=16
```

The comparison focuses on two aspects:

1. Critic and reward-related training curves, to verify that the refactored step-execution path preserves the expected training behavior.
2. `timing_s/gen`, to compare rollout generation performance before and after enabling step execution.

#### Critic and reward curves

The following figure compares the critic and reward-related metrics between the regular full-forward path and the step-execution path.

#### Generation timing

The following figure compares `timing_s/gen` between:

* `step_execution=False`
* `step_execution=True` with `max_num_seqs=16`

### Compatibility

Existing configurations using:

```yaml
actor_rollout_ref:
  rollout:
    step_execution: false
```

continue to use the regular full-forward rollout path.

Configurations using:

```yaml
actor_rollout_ref:
  rollout:
    step_execution: true
    max_num_seqs: 16
```

use the vLLM-Omni step-execution lifecycle through the same `flow_grpo` or `mix_grpo` pipeline adapter.

No algorithm-name change is required.

### Scope

This PR covers:

* Qwen-Image FlowGRPO
* Qwen-Image MixGRPO

Qwen-Image DPO and DiffusionNFT are intentionally excluded from this PR because they use different rollout `custom_output` contracts.

They should be integrated separately instead of being forced to use the FlowGRPO trajectory format.

### Related issue

Implements the first stage of #251.

### CODE REVIEW

File: verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py [hunk 1]
```diff
@@ -54,6 +58,282 @@ def __init__(self, *, od_config: OmniDiffusionConfig, prefix: str = ""):
...
+        self._num_timesteps = len(timesteps)
+
+        if self.transformer.guidance_embeds:
+            guidance = torch.full([1], guidance_scale, dtype=torch.float32)
+            guidance = guidance.expand(latents.shape[0])
+        else:
+            guidance = None
+
+        txt_seq_lens = txt_seq_lens_from_embeds(prompt_embeds)
+        negative_txt_seq_lens = txt_seq_lens_from_embeds(negative_prompt_embeds)
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py:282]: pls fix this

File: verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py [hunk 2]
```diff
@@ -54,6 +58,282 @@ def __init__(self, *, od_config: OmniDiffusionConfig, prefix: str = ""):
...
+        state: "DiffusionRequestState",
+        **kwargs: Any,
+    ) -> "DiffusionRequestState":
+        """Populate *state* with encoded prompts, latents, timesteps, and CFG config.
+
+        Override of ``QwenImagePipeline.prepare_encode`` that accepts pre-tokenized
+        ``prompt_ids`` (and optional ``prompt_mask``) instead of raw text prompts,
+        matching the input contract of ``QwenImagePipelineWithLogProb``.
+        """
+        sampling = state.sampling
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py:195]: pls update the ut, perhaps you can directly modify on tests/workers/rollout/rollout_vllm/test_vllm_omni_generate.py

File: verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sky-Trigger) [reply verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py:282]: The function is:
def txt_seq_lens_from_embeds(prompt_embeds: torch.Tensor | None) -> list[int] | None:
    if prompt_embeds is None:
        return None
    if prompt_embeds.ndim == 2:
        prompt_embeds = prompt_embeds.unsqueeze(0)
    if prompt_embeds.ndim != 3:
        raise ValueError(f"prompt_embeds must be 2D or 3D, got shape={tuple(prompt_embeds.shape)}")
    seq_len = int(prompt_embeds.shape[1])
    batch_size = int(prompt_embeds.shape[0])
    return [seq_len] * batch_size

So this part does not need to be changed.

[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py:282]: I see

File: verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sky-Trigger) [reply verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py:195]: pls check again

**Comment by zhtmike** [comment]
should we update the doc?

File: docs/contributing/common_pitfalls.md
Comment (zhtmike) [inline docs/contributing/common_pitfalls.md:1]: don't touch this

**Comment by Sky-Trigger** [comment]
And now? pls check

**Comment by zhtmike** [comment]
fix CI pls

**Comment by Sky-Trigger** [comment]
> fix CI pls

yes

**Comment by zhtmike** [comment]
still fail

**Comment by Sky-Trigger** [comment]
pls add ci again

**Review by zhtmike — APPROVED** [review]
Thanks for the work!

**Comment by zhtmike** [comment]
@Sky-Trigger can we support bagel as well?

**Comment by Sky-Trigger** [comment]
I took a look, and vLLM-Omni currently does not support stepwise execution for BAGEL. If we want to add BAGEL step execution support in verl-omni, it would be better to add the underlying support in vLLM-Omni first.

**Comment by zhtmike** [comment]
> I took a look, and vLLM-Omni currently does not support stepwise execution for BAGEL. If we want to add BAGEL step execution support in verl-omni, it would be better to add the underlying support in vLLM-Omni first.

I see. Thanks. Btw I got very low gpu utilization at rollout side for bagel. I think it should be very beneficial.
